bottlecap/src/traces/propagation/text_map_propagator.rs

```diff
@@ -0,0 +1,508 @@
+use std::collections::HashMap;
+
+use lazy_static::lazy_static;
+use log::warn;
+use regex::Regex;
+use tracing::{debug, error};
+
+use super::{
+    carrier::{Extractor, Injector},
+    error::Error,
+};
+
+use crate::traces::context::{Sampling, SpanContext};
+
+// Datadog Keys
+const DATADOG_TRACE_ID_KEY: &str = "x-datadog-trace-id";
+const DATADOG_PARENT_ID_KEY: &str = "x-datadog-parent-id";
+const DATADOG_SAMPLING_PRIORITY_KEY: &str = "x-datadog-sampling-priority";
+const DATADOG_ORIGIN_KEY: &str = "x-datadog-origin";
+const DATADOG_TAGS_KEY: &str = "x-datadog-tags";
+
+const DATADOG_HIGHER_ORDER_TRACE_ID_BITS_KEY: &str = "_dd.p.tid";
+const DATADOG_PROPAGATION_ERROR_KEY: &str = "_dd.propagation_error";
+const DATADOG_LAST_PARENT_ID_KEY: &str = "_dd.parent_id";
+const DATADOG_SAMPLING_DECISION_KEY: &str = "_dd.p.dm";
+
+// Traceparent Keys
+const TRACEPARENT_KEY: &str = "traceparent";
+const TRACESTATE_KEY: &str = "tracestate";
+
+lazy_static! {
+    static ref TRACEPARENT_REGEX: Regex =
+        Regex::new(r"(?i)^([a-f0-9]{2})-([a-f0-9]{32})-([a-f0-9]{16})-([a-f0-9]{2})(-.*)?$")
+            .expect("failed creating regex");
+    static ref INVALID_SEGMENT_REGEX: Regex = Regex::new(r"^0+$").expect("failed creating regex");
+    static ref VALID_TAG_KEY_REGEX: Regex =
+        Regex::new(r"^_dd\.p\.[\x21-\x2b\x2d-\x7e]+$").expect("failed creating regex");
+    static ref VALID_TAG_VALUE_REGEX: Regex =
+        Regex::new(r"^[\x20-\x2b\x2d-\x7e]*$").expect("failed creating regex");
+    static ref INVALID_ASCII_CHARACTERS_REGEX: Regex =
+        Regex::new(r"[^\x20-\x7E]+").expect("failed creating regex");
+    static ref VALID_SAMPLING_DECISION_REGEX: Regex =
+        Regex::new(r"^-([0-9])$").expect("failed creating regex");
+}
+
+pub trait TextMapPropagator {
+    fn extract(&self, carrier: &dyn Extractor) -> SpanContext;
+    fn inject(&self, context: SpanContext, carrier: &mut dyn Injector);
+}
+
+#[derive(Clone, Copy)]
+pub struct DatadogPropagator;
+
+impl TextMapPropagator for DatadogPropagator {
+    fn extract(&self, carrier: &dyn Extractor) -> SpanContext {
+        Self::extract_context(carrier).unwrap_or_default()
+    }
+
+    fn inject(&self, _context: SpanContext, _carrier: &mut dyn Injector) {
+        todo!();
```

Comment:
Yeah, this ones i'm not going to particularly adress in this PR, since we don't care about injection yet, maybe I mark them as that they're going to panic since they are not implemented?

Reply:
Nope this is totally fine